url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/commands/model3/Model3Controller.java

```diff
@@ -0,0 +1,725 @@
+				
+				Iterator<GMQueryResult> bIter = this.gm.getNonZeroIterator(this.gm.getRVarWithName(TaskModule.BINDINGNAME), grConds, true);
+				while(bIter.hasNext()){
+					GMQueryResult bRes = bIter.next();
+					
+					String sementicString = getSemanticString((LiftedVarValue)lrRes.getSingleQueryVar(), (LiftedVarValue)bRes.getSingleQueryVar());
+					double jointP = stackedTrajectoryProb * bRes.probability;
+					semSentenceProbs.add(sementicString, jointP);
+					
+				}
+				
+				
+			}
+			
+			
+		}
+		
+		
+		Map<String, Double> semanticStringProbs = new HashMap<String, Double>(semSentenceProbs.size());
+		for(Map.Entry<String, Double> e : semSentenceProbs.entrySet()){
+			double normP = e.getValue() / sumTrajectoryProb;
+			semanticStringProbs.put(e.getKey(), normP);
+		}
+		
+		return semanticStringProbs;
+	}
+	
+	
+	
+	public static String getSemanticString(LiftedVarValue liftedRF, LiftedVarValue bindingConstraints){
+
+		StringBuffer buf = new StringBuffer();
+		boolean first = true;
+		for(GroundedProp gp : liftedRF.conditions){
+			if(!first){
+				buf.append(" ");
+			}
+			buf.append(gp.pf.getName());
+			for(String p : gp.pf.getParameterClasses()){
+				buf.append(" ").append(p);
+			}
+			first = false;
+		}
+		for(GroundedProp gp : bindingConstraints.conditions){
+			buf.append(" ").append(gp.pf.getName());
+			for(String p : gp.pf.getParameterClasses()){
+				buf.append(" ").append(p);
+			}
+		}
+		
+		return buf.toString();
+	}
+	
+	
+	public static List <TrainingElement> getCommandsDataset(Domain domain, String path, StateParser sp){
+		
+		TrainingElementParser teparser = new TrainingElementParser(domain, sp);
+		
+		//get dataset
+		List<TrainingElement> dataset = teparser.getTrainingElementDataset(path, ".txt");
+		
+		return dataset;
+	}
+	
+	public static List<WeightedMTInstance> getStrictMTDataset(Tokenizer tokenizer, String pathToDataDir, String dataFileExtension){
+		
+		//get rid of trailing /
+		if(pathToDataDir.charAt(pathToDataDir.length()-1) == '/'){
+			pathToDataDir = pathToDataDir.substring(0, pathToDataDir.length());
+		}
+		
+		File dir = new File(pathToDataDir);
+		final String ext = new String(dataFileExtension);
+		
+		FilenameFilter filter = new FilenameFilter() {
+			public boolean accept(File dir, String name) {
+				if(name.endsWith(ext)){
+					return true;
+				}
+				return false;
+			}
+		};
+		String[] children = dir.list(filter);
+		
+		List<WeightedMTInstance> dataset = new ArrayList<WeightedMTInstance>(children.length);
+		for(String child : children){
+			String path = pathToDataDir + "/" + child;
+			WeightedMTInstance inst = getStrictMTDataInstanceFromFile(tokenizer, path);
+			dataset.add(inst);
+		}
+		
+		
+		return dataset;
+		
+	}
+	
+	public static WeightedMTInstance getStrictMTDataInstanceFromFile(Tokenizer tokenizer, String pathToFile){
+		
+		WeightedMTInstance d = null;
+		
+		try {
+			BufferedReader in = new BufferedReader(new FileReader(pathToFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 663, 664, 665. The resource is closed at line 670. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).